### PR TITLE
Many Features Added, SimpleOreGenerator Rehaul, Many TE Bug Fixes, And More!

### PR DESCRIPTION
**Pull request by RussDev7**
# Many Features Added, SimpleOreGenerator Rehaul, Many TE Bug Fixes, And More!

## Simple Ore Generator Plugin
This "test" plugin has been fully rehauled and is now fully implimented. Using the plugin will now open up a GUI that has a list of each teir of ore including its counterpart that the user can choose to use.

Example:
TEdit-OreGen

The plugin also now uses a defined list of natual spawn location tiles to check for when generating the ore veins as shown bellow.
```csharp
private bool IsValidLocation(int x, int y)
{
	var curTile = _wvm.CurrentWorld.Tiles[x, y];
    return curTile.IsActive               // Is not Air.
    && curTile.Type == 0                  // Dirt.
    || curTile.Type == 1                  // Stone block.
    || curTile.Type == 25                 // Ebonstone block.
    || curTile.Type == 53                 // Sand block.
    || (curTile.Type == 57 && IncludeAsh) // Ash block.
    || curTile.Type == 59                 // Mud block.
    || curTile.Type == 112                // Ebonsand block.
    || curTile.Type == 116                // Pearlsand block.
    || curTile.Type == 117                // Pearlstone block.
    || curTile.Type == 161                // Ice block.
    || curTile.Type == 163                // Purple ice block.
    || curTile.Type == 164                // Pink ice block.
    || curTile.Type == 200                // Red ice block.
    || curTile.Type == 203                // Crimstone block.
    //    || curTile.Type == 224          // Slush block (Questionable).
    || curTile.Type == 234                // Crimsand block.
    || curTile.Type == 396                // Sandstone block.
    || curTile.Type == 400                // Ebonsandstone block.
    || curTile.Type == 401                // Crimsandstone block.
    || curTile.Type == 403;               // Pearlsandstone block.
}
```
### Conclusion
This now concludes issue #1858.

## New World Height Slider Value Was Broken
The slider for the world height now properly toggles the value it represents. The issue was caused from the portting of TE5. The TE5 now uses `[Reactive]` to ensure UI and other bindings are notified. The intiger `TilesHigh` could not use this tag directly so a simple work around was used.

Before:
TEdit-NewWorld-Broken

After:
TEdit-NewWorld-Fixed

### Conclusion
Issue resolved. Mutiny adverted.

## Can't Place Liquids Behind Solid Blocks
This was something that appeared in TE5. Not too sure why it was introduced but I have reverted it.

Updated Code:
```csharp
// clear liquids for solid tiles
if (curTile.IsActive)
{
	if (WorldConfiguration.TileProperties[curTile.Type].IsSolid &&
        !curTile.InActive &&
        !WorldConfiguration.TileProperties[curTile.Type].IsPlatform &&
        curTile.Type != 52 && // Exclude Vines
        curTile.Type != 62 && // Exclude Jungle Vines
        curTile.Type != 115 && // Exclude Hallowed Vines, 
        curTile.Type != 205 && // Exclude Crimson Vines, 
        curTile.Type != 353 && // Exclude Vine Rope
        curTile.Type != 382 && // Exclude Vine Flowers
        curTile.Type != 365 && // Exclude Silk Rope
        curTile.Type != 366) // Exclude Web Rope
	{
	    // curTile.LiquidAmount = 0;
	}
}
```
### Conclusion
This now concludes issue #1884.

## Add Items To TEdit
The following items where missing from TEdit.

```yaml
ManaCloakStar - 4143
Guide to Critter Companionship (Inactive) - 5453
Guide to Environmental Preservation (Inactive) - 5454
Guide to Peaceful Coexistence (Inactive) - 5455
```
### Conclusion
This now concludes issue #1846.

## Pressing F5 Reloads Current World
By pressing the shortcut key F5 or any programmed key in settings.xml will prompt the user if they want to reload the current world.

Example:
TEdit-Reload

### Conclusion
This now concludes issue #1880.

## The Remove Sprite, Tile, Wall Plugin Crashes On Certian Sprites
The plugin has been updated to ensure that the code handles duplicate entries and prevents out-of-bounds errors.

### Duplicate Key Handling
- Added checks to prevent adding duplicate keys to `tileIds`, `spriteIds`, and `wallIds` dictionaries.
- Logged messages to the console when a duplicate key is detected to aid debugging.

### Bounds Checking
- Retrieved world dimensions using `_wvm.CurrentWorld.TilesWide` and `_wvm.CurrentWorld.TilesHigh`.
- Calculated the range of x and y based on whether a selection is active or if the entire world should be processed.
- Ensured that `x` and `y` indices stay within valid bounds during iteration to avoid index out of bounds errors.

### Iteration Over Tiles
- Used the calculated bounds to safely iterate over the tiles within the specified range.
- Checked and processed tiles, sprite frames, and walls within the valid range only.

### Conclusion
This now concludes two issues: #1820 (From 2022! Woah...), and #1828.

## ReplaceAllPlugin Crashes TEdit If Air Is Selected
As the title suggests, this solves the tile "Air" from causing the plugin to crash.

### Condition Check for Replacing Tiles:
```csharp
if ((_wvm.Selection.IsValid(x, y)) && ((curTile.IsActive && curTile.Type == tileMask && _wvm.TilePicker.TileMaskMode == MaskMode.Match)
    || (!curTile.IsActive && _wvm.TilePicker.TileMaskMode == MaskMode.Empty)
    || (curTile.Type != tileMask && _wvm.TilePicker.TileMaskMode == MaskMode.NotMatching)
    || (tileMask == -1 && !curTile.IsActive))) // Added condition to handle tileMask == -1
{
    doReplaceTile = true;
}
```
- **New Condition:** Added `(tileMask == -1 && !curTile.IsActive)` to handle the case where the `tileMask` is -1 (indicating air). This allows replacement with the `tileTarget` as long as it is not -1.

### Handling tileTarget == -1:
```csharp
if (tileTarget == -1)
{
    // If tileTarget is -1, make the tile inactive
    curTile.IsActive = false;
}
else
{
    curTile.Type = (ushort)tileTarget;
    curTile.IsActive = true;

    if (WorldConfiguration.TileProperties[curTile.Type].IsSolid)
    {
        curTile.U = -1;
        curTile.V = -1;
        BlendRules.ResetUVCache(_wvm, x, y, 1, 1);
    }

    if (_wvm.TilePicker.TilePaintActive)
    {
        curTile.TileColor = (byte)_wvm.TilePicker.TileColor;
    }
}
```
- **New Logic:** If `tileTarget` is -1, the tile is made inactive instead of changing its type. This avoids setting an invalid ID (-1).

### Conclusion
This now concludes issue #1827.

## Chairs Have Incorrect Tile Frames
This issue seemed to be from an error in the `FrameGap`. The old value was `0,4` and it was corrected to `0,2`.

The Silly Balloon Machine was also mentioned to be missing the bottom 3 tiles of its sprite. It's `Size` was incorrect and changed from `3,2` to `3,3`.

### Conclusion
This now concludes issue #1887.

## Allowing Chests To Have Names Longer Then Vanilla 20
This has chaged the vanilla chest name limit from `20` to now `100`. The name property could be longer but it gets to a point of being redundant.

Example:
TEdit-ChestName

### Conclusion
This now concludes issue #1844.

### CODE REVIEW

File: src/TEdit.Terraria/World.Properties.cs [hunk 1]
```diff
@@ -795,15 +795,17 @@ public byte Bg8
     [Reactive] public byte BgTree4 { get; set; }
     [Reactive] public byte UnderworldBg { get; set; }
     [Reactive] public byte MushroomBg { get; set; }
-
-
-
+	
     [Reactive] public int TilesWide { get; set; }
+	[Reactive] public int TilesHighReactive { get; set; }
```
Comment (BinaryConstruct) [inline src/TEdit.Terraria/World.Properties.cs:800]: The seems redundant, but I could be missing something.

File: src/TEdit/View/Popups/NewWorldView.xaml [hunk 2]
```diff
@@ -26,7 +26,7 @@
             </DockPanel>
             <TextBlock Text="Height" HorizontalAlignment="Right" />
             <DockPanel>
-                <TextBlock Text="{Binding TilesHigh, StringFormat={}{0:0}}" Width="30" DockPanel.Dock="Right" />
+                <TextBlock Text="{Binding TilesHighReactive, StringFormat={}{0:0}}" Width="30" DockPanel.Dock="Right" />
```
Comment (BinaryConstruct) [inline src/TEdit/View/Popups/NewWorldView.xaml:29]: This text box and the slider are bound to different properties.

**Comment by BinaryConstruct** [comment]
Note:this was just cursory review, I didn't get a chance to test. Awesome work.

**Comment by BinaryConstruct** [comment]
copilot:walkthrough

File: src/TEdit.Terraria/World.Properties.cs (hunk 1 above)
Comment (RussDev7) [reply src/TEdit.Terraria/World.Properties.cs:800]: Due to how `[Reactive]` works, and being it's get and set control members are declared in another module, this was the only feasible workaround to this I personally could find without it getting crazy.

File: src/TEdit/View/Popups/NewWorldView.xaml (hunk 2 above)
Comment (RussDev7) [reply src/TEdit/View/Popups/NewWorldView.xaml:29]: Yes this is intended. This ensures that `TilesHigh` still controls its background UI and also updates `UpdateMaxLayerLevels();`. The slider is supposed to be controlling `TilesHigh` while its reactive is updating and reflecting back `TilesHighReactive`. This solves the issues mentioned above.

**Pull request merged**